File: src/main/java/com/example/hbs/psi/impl/HbOpenBlockMustacheImpl.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.example.hbs.psi.impl;

import com.example.hbs.psi.HbCloseBlockMustache;
import com.example.hbs.psi.HbOpenBlockMustache;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import icons.HandlebarsIcons;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

public class HbOpenBlockMustacheImpl extends HbBlockMustacheImpl implements HbOpenBlockMustache {
  public HbOpenBlockMustacheImpl(@NotNull ASTNode astNode) {
    super(astNode);
  }

  @Override
  public HbCloseBlockMustache getPairedElement() {
    PsiElement closeBlockElement = getParent().getLastChild();
    if (closeBlockElement instanceof HbCloseBlockMustache) {
      return (HbCloseBlockMustache)closeBlockElement;
    }

    return null;
  }

  @Override
  public @Nullable Icon getIcon(int flags) {
    return HandlebarsIcons.Elements.OpenBlock;
  }
}

File: src/main/kotlin/com/example/hbs/psi/impl/HbOpenBlockMustacheImpl.kt
package com.example.hbs.psi.impl

import com.example.hbs.psi.HbCloseBlockMustache
import com.example.hbs.psi.HbOpenBlockMustache
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiElement
import icons.HandlebarsIcons
import javax.swing.Icon

class HbOpenBlockMustacheImpl(astNode: ASTNode) :
    HbBlockMustacheImpl(astNode), HbOpenBlockMustache {

    override fun getPairedElement(): HbCloseBlockMustache {
        // Идём по всем детям родителя в обратном порядке и ищем HbCloseBlockMustache
        var element: PsiElement? = parent.lastChild
        while (element != null) {
            if (element is HbCloseBlockMustache) return element
            element = element.prevSibling
        }
        // Если не нашли — выбрасываем исключение, так как метод не nullable
        throw IllegalStateException("Paired close block not found for $this")
    }

    override fun getIcon(flags: Int): Icon? {
        return HandlebarsIcons.Elements.OpenBlock
    }
}
